utilities/env_librados.cc

```diff
@@ -0,0 +1,961 @@
+}
+
+rocksdb::Status EnvLibrados::CreateDirIfMissing(const std::string& dirname)
+{
+    std::string fid, dir, file;
+    split(dirname+"/temp", &dir, &file);
+    rocksdb::Status s = _GetFid(dir+"/"+file, fid);
+    if (rocksdb::Status::NotFound() != s) {
+        return rocksdb::Status::IOError();
+    }
+
+    return _AddFid(dir+"/"+file, DIR_TAG_VALUE);
+}
+
+/* Only delete empty dir; otherwise return error */
+rocksdb::Status EnvLibrados::DeleteDir(const std::string& dirname)
+{
+  std::string fid, dir, file;
+  split(dirname+"/temp", &dir, &file);
+  rocksdb::Status s = _GetFid(dir, fid);
+  
+  if (s.ok() && DIR_TAG_VALUE == fid) {
+      return _DelFid(dir);
+  } else {
+      return rocksdb::Status::NotFound();
+  }
+}
+
+rocksdb::Status EnvLibrados::GetFileSize(
+  const std::string& fname,
+  uint64_t* file_size)
+{
+    std::string fid, dir, file;
+    split(fname, &dir, &file);
+    time_t mtime;
+    rocksdb::Status s = _GetFid(dir+"/"+file, fid);
+
+    if (!s.ok()) {
+        return s;
+    }
+
+    int ret = _io_ctx.stat(fid, file_size, &mtime);
+    LOG_DEBUG(" %s|%lld|%i\n", fid.c_str(), static_cast<long long>(*file_size), ret);
+    if (ret < 0) {
+        if (rocksdb::Status::NotFound() == err_to_status(ret)) {
+            *file_size = 0;
+            return rocksdb::Status::OK();
+        } else {
+            return err_to_status(ret);
+        }
+    } else {
+        return rocksdb::Status::OK();
+    }
+}
+
+rocksdb::Status EnvLibrados::GetFileModificationTime(const std::string& fname,
+                    uint64_t* file_mtime)
+{
+    std::string fid, dir, file;
+    split(fname, &dir, &file);
+    time_t mtime;
+    uint64_t file_size;
+    rocksdb::Status s = _GetFid(dir+"/"+file, fid);
+
+    if (!s.ok()) {
+        return s;
+    }
+
+    int ret = _io_ctx.stat(fid, &file_size, &mtime);
+    LOG_DEBUG(" %s|%lld|%i\n", fid.c_str(), static_cast<long long>(mtime), ret);
+    if (ret < 0) {
+        if (rocksdb::Status::NotFound() == err_to_status(ret)) {
+            *file_mtime = static_cast<uint64_t>(mtime);
+            return rocksdb::Status::OK();
+        } else {
+            return err_to_status(ret);
+        }
+    } else {
+        return rocksdb::Status::OK();
+    }
+
+    return rocksdb::Status::OK();
+}
+
+
+/* This is not atomic operation. First delete old file, then create new file */
+rocksdb::Status EnvLibrados::RenameFile(
+  const std::string& src,
+  const std::string& target_in)
+{
+    std::string src_fid, tmp_fid, src_dir, src_file, dst_dir, dst_file;
+    split(src, &src_dir, &src_file);
+    split(target_in, &dst_dir, &dst_file);
+    
+    auto s = _GetFid(src_dir + "/" + src_file, src_fid);
+    if (!s.ok()) {
+        return s;
+    }
+    
+    s = _GetFid(dst_dir + "/" + dst_file, tmp_fid);
+    if (s.ok()) {
+        return rocksdb::Status::IOError();
+    }
+    
+    s = _DelFid(src_dir + "/" + src_file);
+    if (!s.ok()) {
+        return rocksdb::Status::IOError();
+    }
```

Comment:
Let's make this atomic.  _RenameFid(oldfid, newfid)

The way tod o this is 
<pre>
librados::ObjectWriteOperation o;
o.omap_rmkeys(...);
o.omap_set(...);
_io_ctx.operate(o, ...);
</pre>